components/script/dom/mediafragmentparser.rs

```diff
@@ -0,0 +1,355 @@
+                    &input.replace('-', "").replace(',', ""),
+                    "%Y%m%dT%H%M%S%.fZ",
+                )
+                .map_err(|_| ())?
+                .time()
+                .to_string()
+                .as_ref(),
+            )?;
+            if input.starts_with(',') {
+                Ok((Some(0.), Some(sec)))
+            } else {
+                Ok((Some(sec), None))
+            }
+        } else {
+            let vec: Vec<&str> = input.split('-').collect();
+            let mut hms: Vec<f64> = vec
+                .iter()
+                .map(|s| NaiveDateTime::parse_from_str(s, "%Y%m%dT%H%M%S%.fZ"))
+                .flatten()
+                .map(|s| parse_hms(&s.time().to_string()))
+                .flatten()
+                .collect();
+
+            let end = hms.pop().ok_or_else(|| ())?;
+            let start = hms.pop().ok_or_else(|| ())?;
+
+            if !hms.is_empty() || start >= end {
+                return Err(());
+            }
+
+            Ok((Some(start), Some(end)))
+        }
+    }
+
+    fn parse_spatial(&self, input: &str) -> Result<SpatialClipping, ()> {
+        let (prefix, s) = split_prefix(input);
+        let vec: Vec<&str> = s.split(',').collect();
+        let mut queue: VecDeque<u32> = vec.iter().map(|s| s.parse::<u32>()).flatten().collect();
+
+        let mut clipping = SpatialClipping {
+            region: None,
+            x: queue.pop_front().ok_or_else(|| ())?,
+            y: queue.pop_front().ok_or_else(|| ())?,
+            width: queue.pop_front().ok_or_else(|| ())?,
+            height: queue.pop_front().ok_or_else(|| ())?,
+        };
+
+        if !queue.is_empty() {
+            return Err(());
+        }
+
+        if let Some(s) = prefix {
+            let region = SpatialRegion::from_str(s)?;
+            if region.eq(&SpatialRegion::Percent) &&
+                (clipping.x + clipping.width > 100 || clipping.y + clipping.height > 100)
+            {
+                return Err(());
+            }
+            clipping.region = Some(region);
+        }
+
+        Ok(clipping)
+    }
+}
+
+impl From<&Url> for MediaFragmentParser {
+    fn from(url: &Url) -> Self {
+        let input: &str = &url[Position::AfterPath..];
+        MediaFragmentParser::parse(input)
+    }
+}
+
+impl From<&ServoUrl> for MediaFragmentParser {
+    fn from(servo_url: &ServoUrl) -> Self {
+        let input: &str = &servo_url[Position::AfterPath..];
+        MediaFragmentParser::parse(input)
+    }
+}
+
+// 5.1.1 Processing name-value components.
+fn decode_octets(bytes: &[u8]) -> Vec<(Cow<str>, Cow<str>)> {
+    form_urlencoded::parse(bytes)
+        .filter(|(key, _)| match key.as_bytes() {
+            b"t" | b"track" | b"id" | b"xywh" => true,
+            _ => false,
+        })
+        .collect()
+}
+
+// Parse a full URL or a relative URL without a base retaining the query and/or fragment.
+fn split_url(s: &str) -> (DOMString, DOMString) {
```

Comment:
Shouldn't this and other parts of this module be using the `url` crate?

Reply:
This is mostly for relative URL's which could be passed to the parser from HTMLMediaElement for the video/audio src. These may have a relative file location with a query or fragment such as: resources/video.mp4#t=10